sdk/storage/azure-storage-stress/src/main/java/com/azure/storage/stress/CrcInputStream.java

```diff
@@ -6,48 +6,83 @@
+import java.util.Arrays;
 import java.util.zip.CRC32;
 
 public class CrcInputStream extends InputStream {
     private final static ClientLogger LOGGER = new ClientLogger(CrcInputStream.class);
-    private final RepeatingInputStream repeatingInputStream;
+    private final Sinks.One<ContentInfo> sink = Sinks.one();
+    private final InputStream inputStream;
     private final CRC32 crc = new CRC32();
+    private final byte[] head = new byte[1024];
+    private long length = 0;
 
     public CrcInputStream(BinaryData source, long size) {
-        this.repeatingInputStream = new RepeatingInputStream(source, size);
+        this.inputStream = new RepeatingInputStream(source, size);
+    }
+
+    public CrcInputStream(InputStream source) {
+        this.inputStream = source;
     }
 
     @Override
-    public synchronized int read() {
-        int b = repeatingInputStream.read();
-        if (b >= 0) {
-            crc.update(b);
+    public synchronized int read() throws IOException {
+        try {
+            int b = inputStream.read();
+            if (b >= 0) {
+                crc.update(b);
+                if (length < head.length) {
+                    head[(int) length] = (byte) b;
+                }
+                length++;
+            }
+            if (b == -1) {
+                sink.emitValue(new ContentInfo(crc.getValue(), length, head), Sinks.EmitFailureHandler.FAIL_FAST);
+            }
+            return b;
+        } catch (IOException e) {
+            sink.emitError(e, Sinks.EmitFailureHandler.FAIL_FAST);
+            throw LOGGER.logThrowableAsError(e);
         }
-
-        return b;
     }
 
     @Override
-    public synchronized int read(byte b[], int off, int len) {
-        int read = repeatingInputStream.read(b, off, len);
-        if (read > 0) {
-            crc.update(b, off, read);
+    public synchronized int read(byte b[], int off, int len) throws IOException {
+        try {
+            int read = inputStream.read(b, off, len);
+            if (read > 0) {
+                crc.update(b, off, read);
+                if (length < head.length) {
+                    System.arraycopy(b, off, head, (int)length, Math.min(read, head.length - (int)length));
+                }
+                length += read;
+            }
+            if (read == -1) {
+                sink.emitValue(new ContentInfo(crc.getValue(), length, head), Sinks.EmitFailureHandler.FAIL_FAST);
+            }
+            return read;
+        } catch (IOException e) {
+            sink.emitError(e, Sinks.EmitFailureHandler.FAIL_FAST);
+            throw LOGGER.logThrowableAsError(e);
         }
-        return read;
     }
 
-    public synchronized long getCrc() {
-        return crc.getValue();
+    public Mono<ContentInfo> getContentInfo() {
+        return sink.asMono();
     }
 
     @Override
     public void close() {
         try {
-            repeatingInputStream.close();
+            inputStream.close();
```

Comment:
Isn't `close` checked with IOException by default? Do we need the try-catch here?

Reply:
it makes using it easier, e.g. I can do 

```java
        return Mono.using(
                () -> new CrcInputStream(BLOB_CONTENT_HEAD, blobSize),
                data -> ..
                CrcInputStream::close)
```

If close throws, I'd have to write more boilerplate.